Allow loading client from env

jwygoda opened the pull request.

BenRKarl commented:
@jwygoda Thanks for posting this! Would you please add unit tests for the client for these new methods? There should be some existing examples to build off of in the `tests/client_tests.py` file.

jwygoda commented:
No problem. I've added tests for new methods and moved some of them from load_from_storage to _get_client_kwargs. I've also splitted test_load_from_storage_too_short_login_customer_id and test_load_from_storage_invalid_login_customer_id into test_validate_login_customer_id_invalid, test_validate_login_customer_id_too_short and test_init_validate_login_customer_id. Hope that's ok with you.

BenRKarl commented on google/ads/google_ads/client.py:
```diff
@@ -38,16 +38,22 @@
 
         Args:
-            yaml_str: a str containing client configuration in YAML format.
+            config_data: a dict containing client configuration.
```
Can you add a comment for `error_message`

jwygoda replied:
Fixed

BenRKarl commented on README.rst:
```diff
@@ -46,13 +46,45 @@ is a `template`_ you can use.
+
+  export GOOGLE_ADS_LOGIN_CUSTOMER_ID=INSERT_LOGIN_CUSTOMER_ID_HERE
+  export GOOGLE_ADS_ENDPOINT=INSERT_ALTERNATIVE_API_ENDPOINT_HERE
+  export GOOGLE_ADS_LOGGING_CONFIG=INSERT_LOGGING_CONFIG_HERE
```
What would the logging config look like coming from a bash env var? Do you think you could provide an example here that uses the same configuration that's in the sample `google-ads.yaml` file so that it's easy to compare?

jwygoda replied:
Added example to readme

BenRKarl commented:
@jwygoda Heads up I may merge this and submit additional documentation in a separate PR so that this functionality doesn't get held up. Thanks again for the PR it's very much appreciated!

wihl commented on README.rst:
```diff
@@ -47,13 +47,54 @@ is a `template`_ you can use.
 Create a GoogleAdsClient
 ########################
 
+Using yaml file
```
s/yaml/YAML

BenRKarl replied:
A note here that `s/yaml/YAML` may not be clear, and means: _replace `yaml` with `YAML`_.

I don't think it's worth holding this PR up for these changes. I can make them in a later release.

jwygoda replied:
Fixed

wihl commented on README.rst:
```diff
@@ -47,13 +47,54 @@ is a `template`_ you can use.
+.. code-block:: bash
+
+  export GOOGLE_ADS_LOGIN_CUSTOMER_ID=INSERT_LOGIN_CUSTOMER_ID_HERE
+  export GOOGLE_ADS_ENDPOINT=INSERT_ALTERNATIVE_API_ENDPOINT_HERE
```
I don't think this should go in the README as this is really for internal Google testing and would never be used by external users.

jwygoda replied:
Fixed

wihl commented on README.rst:
```diff
@@ -47,13 +47,54 @@ is a `template`_ you can use.
+
+.. _GOOGLE_ADS_LOGGING:
+
+GOOGLE_ADS_LOGGING should be a json with logging configuration. Example:
```
s/json/JSON string/

jwygoda replied:
Fixed

wihl commented on README.rst:
```diff
@@ -91,8 +132,9 @@ The currently available list of versions is:
-configuration via the configuration file; see `google-ads.yaml`_
-for an example. The library logs to ``stderr`` by default. You can easily pipe
+configuration via the configuration file (see `google-ads.yaml`_
+for an example) or GOOGLE_ADS_LOGGING_ environment variable.
```
extra space after GOOGLE_ADS_LOGGING

wihl commented on google/ads/google_ads/client.py:
```diff
@@ -38,16 +38,24 @@
+_ENV_PREFIX = 'GOOGLE_ADS_'
+_KEYS_ENV_VARIABLES_MAP = {
+    key: _ENV_PREFIX + key.upper() for key in
+    list(_REQUIRED_KEYS) + ['login_customer_id', 'endpoint', 'logging']
```
again, remove endpoint

BenRKarl replied:
We probably want to keep endpoint here, otherwise if we're configuring the lib with env vars we can't set the endpoint. I think it's reasonable not to document, but there should be symmetry with the k/vs you can set in each mode of configuration, i.e. it doesn't make sense that endpoint can be set via yaml but not via an env var.

jwygoda replied:
Removed endpoint from env vars as it was originally requested. Should I restore it?

BenRKarl replied:
Yes please, sorry about the confusion.

jwygoda replied:
Done

wihl commented on google/ads/google_ads/client.py:
```diff
@@ -77,9 +83,64 @@ def _get_client_kwargs_from_yaml(cls, yaml_str):
+
+        Raises:
+            ValueError: If the environment lacks a required field or
+                GOOGLE_ADS_LOGGING env variable is not in json format.
```
nit: s/json/JSON

jwygoda replied:
Fixed

wihl commented on google/ads/google_ads/client.py:
```diff
@@ -77,9 +83,64 @@ def _get_client_kwargs_from_yaml(cls, yaml_str):
+                config_data['logging'] = json.loads(config_data['logging'])
+            except json.JSONDecodeError:
+                raise ValueError(
+                    'GOOGLE_ADS_LOGGING env variable should be in json format.'
```
nit: s/json/JSON

jwygoda replied:
Fixed

wihl commented on README.rst:
```diff
@@ -133,7 +132,7 @@ Enabling and Configuring logging
 The library uses Python's built in logging framework. You can specify your
 configuration via the configuration file (see `google-ads.yaml`_
-for an example) or GOOGLE_ADS_LOGGING_ environment variable.
+for an example) or GOOGLE_ADS_LOGGING_  environment variable.
```
There was an extra space - now there are two. Not a big deal

jwygoda replied:
Fixed
